src/pybind/mgr/dashboard/controllers/auth.py

```diff
@@ -28,33 +29,48 @@ class Auth(RESTController):
-            pwd_expiration_date = user_data.get('pwdExpirationDate', None)
-            pwd_update_required = user_data.get('pwdUpdateRequired', False)
-
-        if user_perms is not None:
-            logger.debug('Login successful')
-            token = JwtManager.gen_token(username)
-            token = token.decode('utf-8')
-            cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
-            return {
-                'token': token,
-                'username': username,
-                'permissions': user_perms,
-                'pwdExpirationDate': pwd_expiration_date,
-                'sso': mgr.SSO_DB.protocol == 'saml2',
-                'pwdUpdateRequired': pwd_update_required
-            }
+        max_attempt = Settings.ACCOUNT_LOCKOUT_ATTEMPTS
+        user_attempts = mgr.ACCESS_CTRL_DB.get_attempt(username)
+        if max_attempt == 0 or user_attempts < max_attempt:
+            if user_data:
+                user_perms = user_data.get('permissions')
+                pwd_expiration_date = user_data.get('pwdExpirationDate', None)
+                pwd_update_required = user_data.get('pwdUpdateRequired', False)
 
-        logger.debug('Login failed')
-        raise DashboardException(msg='Invalid credentials',
-                                 code='invalid_credentials',
-                                 component='auth')
+            if user_perms is not None:
+                logger.info('Login successful: %s', username)
+                mgr.ACCESS_CTRL_DB.save_attempt(username, 0)
+                mgr.ACCESS_CTRL_DB.save()
+                token = JwtManager.gen_token(username)
+                token = token.decode('utf-8')
+                cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
+                return {
+                    'token': token,
+                    'username': username,
+                    'permissions': user_perms,
+                    'pwdExpirationDate': pwd_expiration_date,
+                    'sso': mgr.SSO_DB.protocol == 'saml2',
+                    'pwdUpdateRequired': pwd_update_required
+                }
+            user_attempts += 1
+            mgr.ACCESS_CTRL_DB.save_attempt(username, user_attempts)
+            mgr.ACCESS_CTRL_DB.save()
+        else:
+            try:
+                user = mgr.ACCESS_CTRL_DB.get_user(username)
+                user.enabled = False
+                mgr.ACCESS_CTRL_DB.save()
+                logging.warning("""Maximum number of unsuccessful log-in attempts (%d) reached
+                                for username '%s', so the account was blocked. An administrator will
+                                need to re-enable the account""", max_attempt, username)
+                raise InvalidCredentialsError
+            except UserDoesNotExist:
+                raise InvalidCredentialsError
+        logger.info('Login failed: %s', username)
+        raise InvalidCredentialsError
```

Comment:
```suggestion
                raise InvalidCredentialsError()
            except UserDoesNotExist:
                raise InvalidCredentialsError()
        logger.info('Login failed: %s', username)
        raise InvalidCredentialsError()
```

This is optional, I would suggest creating an instance.

Is it possible to let the user know if his account is locked? Even a user account is locked, he still gets a toast about the invalid credential:

![Screen Shot 2020-12-03 at 3 22 53 PM](https://user-images.githubusercontent.com/1691518/100977144-83000c00-357b-11eb-9a6c-b8d355a9a863.png)

Something more meaningful will be better.

Reply:
@bk201 I had the similar opinion about this and here is @epuertat opinion on it https://github.com/ceph/ceph/pull/38316#issuecomment-737145178

Reply:
Understand, I missed that conversation. Let's stick with what you have.